# Exercises I
#### Run python -m pip install pymysql from your terminal to install pymysql.

##### Create a notebook or python script named advanced_dataframes to do your work in for these exercises.

1. Run python -m pip install pymysql from your terminal to install the mysql client (any folder is fine)

2. cd into your exercises folder for this module and run echo env.py >> .gitignore

3. Create a function named get_db_url. It should accept a username, hostname, password, and database name and return a url connection string formatted like in the example at the start of this lesson.

4. Use your function to obtain a connection to the employees database.

5. Once you have successfully run a query:

   a. Intentionally make a typo in the database url. What kind of error message do you see?

   b. Intentionally make an error in your SQL query. What does the error message look like?

6. Read the employees and titles tables into two separate DataFrames.

7. How many rows and columns do you have in each DataFrame? Is that what you expected?

8. Display the summary statistics for each DataFrame.

9. How many unique titles are in the titles DataFrame?

10. What is the oldest date in the to_date column?

11. What is the most recent date in the to_date column?

## Exercises I

Run `python -m pip install mysqlclient pymysql` from your terminal to install pymysql and the mysqlclient.

Create a notebook or python script named `advanced_dataframes` to do your work in for these exercises.

1. Create a function named `get_db_url`. It should accept a username, hostname, password, and database name and return a url connection string formatted like in the example at the start of this lesson.

In [28]:
# def get_db_url(schema):
#     import env
#     user = env.username
#     password = env.password
#     host = env.host
#     conn = f'mysql+pymysql://{user}:{password}@{host}/{schema}'
#     return conn

In [ ]:
def get_db_url(db, env_file=os.path.exists('env.py')):
    '''
    returns a formatted string ready to utilize as a sql url connection
    
    args: db: a string literal representing a schema
    env_file: bool: checks to see if there is an env.py present in the cwd
    
    make sure that if you have an env file that you import it outside of the scope 
    of this function call, otherwise env.user wont mean anything ;)
    '''
    import os
    if env_file:
        user, password, host = (env.user, env.password, env.host)
        return f'mysql+pymysql://{user}:{password}@{host}/{db}'
    else:
        return 'yo you need some credentials to access a database usually and I dont want you to type them here.'

2. Use your function to obtain a connection to the `employees` database.

In [31]:
import numpy as np
import pandas as pd

In [33]:
query = 'SELECT * FROM employees LIMIT 100'

In [34]:
df = pd.read_sql(query, get_db_url('employees'))

In [35]:
df

emp_no  birth_date   first_name    last_name gender   hire_date
0    10001  1953-09-02       Georgi      Facello      M  1986-06-26
1    10002  1964-06-02      Bezalel       Simmel      F  1985-11-21
2    10003  1959-12-03        Parto      Bamford      M  1986-08-28
3    10004  1954-05-01    Chirstian      Koblick      M  1986-12-01
4    10005  1955-01-21      Kyoichi     Maliniak      M  1989-09-12
..     ...         ...          ...          ...    ...         ...
95   10096  1954-09-16       Jayson      Mandell      M  1990-01-14
96   10097  1952-02-27        Remzi  Waschkowski      M  1990-09-15
97   10098  1961-09-23  Sreekrishna    Servieres      F  1985-05-13
98   10099  1956-05-25       Valter      Sullins      F  1988-10-18
99   10100  1953-04-21     Hironobu    Haraldson      F  1987-09-21

[100 rows x 6 columns]

3. Once you have successfully run a query:

    a. Intentionally make a typo in the database url. What kind of error message do you see?
    

In [36]:
ten_emp_no = pd.read_sql('SELECT emp_no FROM employees LIMIT 10', get_db_url('employeeees') )

OperationalError: (pymysql.err.OperationalError) (1044, "Access denied for user 'leavitt_1877'@'%' to database 'employeeees'")
(Background on this error at: https://sqlalche.me/e/14/e3q8)

### OperationalError: The Connection Fairy didn't like me

    b. Intentionally make an error in your SQL query. What does the error message look like?

In [37]:
ten_emp_no = pd.read_sql('SEECT emp_no FROM employees LIMIT 10', get_db_url('employees') )

ProgrammingError: (pymysql.err.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'SEECT emp_no FROM employees LIMIT 10' at line 1")
[SQL: SEECT emp_no FROM employees LIMIT 10]
(Background on this error at: https://sqlalche.me/e/14/f405)

#### ProgrammingError: 1064, You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version

In [38]:
ten_emp_no = pd.read_sql('SELECT emp_no FROM employees LIMIT 10', get_db_url('employees') )

In [40]:
ten_emp_no

emp_no
0   10001
1   10002
2   10003
3   10004
4   10005
5   10006
6   10007
7   10008
8   10009
9   10010

4. Read the `employees` and `titles` tables into two separate DataFrames.

In [41]:
employees = pd.read_sql('SELECT * FROM employees', get_db_url('employees'))

In [43]:
titles = pd.read_sql('SELECT * FROM titles', get_db_url('employees'))

5. How many rows and columns do you have in each DataFrame? Is that what you expected?

In [44]:
employees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300024 entries, 0 to 300023
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   emp_no      300024 non-null  int64 
 1   birth_date  300024 non-null  object
 2   first_name  300024 non-null  object
 3   last_name   300024 non-null  object
 4   gender      300024 non-null  object
 5   hire_date   300024 non-null  object
dtypes: int64(1), object(5)
memory usage: 13.7+ MB


In [45]:
titles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 443308 entries, 0 to 443307
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   emp_no     443308 non-null  int64 
 1   title      443308 non-null  object
 2   from_date  443308 non-null  object
 3   to_date    443308 non-null  object
dtypes: int64(1), object(3)
memory usage: 13.5+ MB


6. Display the summary statistics for each DataFrame.

In [46]:
employees.describe()

emp_no
count  300024.000000
mean   253321.763392
std    161828.235540
min     10001.000000
25%     85006.750000
50%    249987.500000
75%    424993.250000
max    499999.000000

In [65]:
employees.to_csv('employees_employees.csv')

In [66]:
titles.to_csv('employees_titles.csv')

In [47]:
titles.describe()

emp_no
count  443308.000000
mean   253075.034430
std    161853.292613
min     10001.000000
25%     84855.750000
50%    249847.500000
75%    424891.250000
max    499999.000000

7. How many unique titles are in the `titles` DataFrame?

In [51]:
titles.title.unique(), len(titles.title.unique())

(array(['Senior Engineer', 'Staff', 'Engineer', 'Senior Staff',
        'Assistant Engineer', 'Technique Leader', 'Manager'], dtype=object),
 7)

8. What is the oldest date in the `to_date` column? 

In [52]:
titles.to_date.min()

datetime.date(1985, 3, 1)

9. What is the most recent date in the `to_date` column?

In [58]:
titles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 443308 entries, 0 to 443307
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   emp_no     443308 non-null  int64 
 1   title      443308 non-null  object
 2   from_date  443308 non-null  object
 3   to_date    443308 non-null  object
dtypes: int64(1), object(3)
memory usage: 13.5+ MB


In [60]:
titles['from_date'] = pd.to_datetime(titles['from_date'])

In [62]:
titles['to_date'].head()

0    9999-01-01
1    9999-01-01
2    9999-01-01
3    1995-12-01
4    9999-01-01
Name: to_date, dtype: object

In [69]:
import datetime as dt


In [75]:
help()


Welcome to Python 3.9's help utility!

If this is your first time using Python, you should definitely check out
the tutorial on the Internet at https://docs.python.org/3.9/tutorial/.

Enter the name of any module, keyword, or topic to get help on writing
Python programs and using Python modules.  To quit this help utility and
return to the interpreter, just type "quit".

To get a list of available modules, keywords, symbols, or topics, type
"modules", "keywords", "symbols", or "topics".  Each module also comes
with a one-line summary of what it does; to list the modules whose name
or summary contain a given string such as "spam", type "modules spam".

help> dt
No Python documentation found for 'dt'.
Use help() to get the interactive help utility.
Use help(str) for help on the str class.

help> datetime
Help on module datetime:

NAME
    datetime - Fast implementation of the datetime type.

MODULE REFERENCE
    https://docs.python.org/3.9/library/datetime
    
    The following documen


You are now leaving help and returning to the Python interpreter.
If you want to ask for help on a particular object directly from the
interpreter, you can type "help(object)".  Executing "help('string')"
has the same effect as typing a particular string at the help> prompt.


In [79]:
titles['to_date'][titles['to_date']!=dt.date(9999,1,1,)].max()

datetime.date(2002, 8, 1)

In [61]:
titles['to_date'] = pd.to_datetime(titles['to_date'])

OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 9999-01-01 00:00:00

In [57]:
titles.to_date[titles.to_date < pd.to_datetime(['9999-01-01'])].max()

OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 9999-01-01 00:00:00

In [86]:
print(titles['to_date'][titles['to_date']!=dt.date(9999,1,1)].max())

2002-08-01
